### PR TITLE
List and dict improvements

### PR DESCRIPTION
**Pull request by freakboy3742**
This improves handling for the bridging of Python lists and dictionaries, including the use of primitive values as keys and values.

### CODE REVIEW

**Comment by dgelessus** [comment]
I like the automatic boxing of primitive values. This allows automatic conversion of lists/dicts containing Python primitives, right? (e. g. Python `{"zero": 0, "one": 1, "two": 2}` is converted to Objective-C `@{@"zero": @0, @"one": @1, @"two": @2}`)

I'm not sure about the custom `__repr__` and `__str__` for the Objective-C collections. `ObjCInstance` already provides custom `__repr__` and `__str__` methods, which call the Objective-C `debugDescription` method. (`__str__` returns the `debugDescription` unmodified, `__repr__` includes some additional info about the Python and Objective-C classes and addresses.) The `debugDescription` for arrays and dictionaries already provides a good string representation IMHO, so the custom `__repr__` and `__str__` seem a bit redundant to me.

**Comment by freakboy3742** [comment]
Yes - the autoboxing of primitives means you can pass in almost any Python list, dict, or primitive and it (or the things it contains) will get converted to "object form".

As for `__str__` and `__repr__` - Those came out of the debugging process for the auto boxing :-) Yes - the Objective C objects fall back onto debugDescription et al, but the output format leaves a lot to be desired. It tells you the memory address of a list, but not the things that are *in* the list. And OK, sure - you can easily find that out and print it, but it doesn't make for a very Pythonic experience.

If you think it's worth preserving memory details in the __repr__, I can definitely see the benefit - but in Python space, `__str__` really should be a "useful" output, and the `debugDescription` definitely isn't that.

**Comment by dgelessus** [comment]
That's odd, I have no issues with the default `__repr__` and `__str__` showing me the array contents:

```python
>>> NSMutableArray = objc.ObjCClass("NSMutableArray")
>>> arr = NSMutableArray.array()
>>> arr.append(objc.at("spam"))
>>> arr.append(objc.at("ham"))
>>> arr.append(objc.at("eggs"))
>>> arr
Out[9]: 
<rubicon.objc.runtime.ObjCMutableListInstance 0x104be3e10: __NSArrayM at 0x10076cba0: <__NSArrayM 0x10076cba0>(
spam,
ham,
eggs
)
>
>>> print(str(arr))
<__NSArrayM 0x10076cba0>(
spam,
ham,
eggs
)
```

To be honest, the default `__repr__` is a bit long, maybe we should shorten that. For example I don't think the Python `id` matters in practice, and we can probably leave out the Objective-C object's address as well (if it matters, the `debugDescription` probably includes it anyway).

**Comment by freakboy3742** [comment]
For me, the `__repr__` is probably OK; it's verbose, yes - but then it's a repr - it's meant to be a peek into internals. But `__str__` is meant to be human readable - something you could put in a user interface. The format currently being used doesn't meet that requirement.

**Comment by dgelessus** [comment]
Should we change the generic `__str__` to call `description` instead of `debugDescription`? The `description` is usually a bit less verbose, for example for arrays it doesn't include the class name and address:

```python
print(arr.description)
(
    spam,
    ham,
    eggs
)
```

**Comment by freakboy3742** [comment]
That sounds like it might work. I'll make that adjustment.

File: rubicon/objc/core_foundation.py [hunk 1]
```diff
@@ -67,8 +67,9 @@
 
 # Core Foundation type to Python type conversion functions
 def CFSTR(string):
-    return ObjCInstance(libcf.CFStringCreateWithCString(
-            None, string.encode('utf-8'), kCFStringEncodingUTF8))
+    return ObjCInstance(cf.CFStringCreateWithCString(
```
Comment (dgelessus) [inline rubicon/objc/core_foundation.py]: That should be `libcf` instead of `cf` (the change got lost in the merge).

File: tests/test_NSArray.py [hunk 2]
```diff
@@ -25,8 +25,6 @@
 
 class NSArrayMixinTest(unittest.TestCase):
     py_list = ['one', 'two', 'three']
-    name = 'NSArray'
-
     def make_array(self, contents=None):
```
Comment (dgelessus) [inline tests/test_NSArray.py]: Missing blank line, `flake8` does not like that.

File: tests/test_NSArray.py [hunk 3]
```diff
@@ -23,7 +25,6 @@
 
 class NSArrayMixinTest(unittest.TestCase):
     py_list = ['one', 'two', 'three']
-
     def make_array(self, contents=None):
```
Comment (dgelessus) [inline tests/test_NSArray.py]: Missing blank line, `flake8` does not like that.

**Comment by dgelessus** [comment]
The Travis log also shows [a segfault](https://travis-ci.org/pybee/rubicon-objc/jobs/276231034#L1707), but that might just be a consequence of the `NameError` right before.

**Pull request merged**